src/sage/modules/diamond_cutting.py

```diff
@@ -259,19 +269,88 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
         1
+
+    TESTS:
+
+    Verify that :issue:`39507` is fixed::
+
+        sage: from sage.modules.free_module_integer import IntegerLattice
+        sage: v = vector(ZZ, [1,1,1,-1])
+        sage: L = IntegerLattice([v])
+        sage: print(v in L)
+        True
+        sage: print(L.closest_vector(v))
+        (1, 1, 1, -1)
+        sage: v = vector(ZZ, [1,1,1,-1])
+        sage: L = IntegerLattice([v])
+        sage: C = L.voronoi_cell()
+        sage: C.Hrepresentation()
+        (An inequality (-1, -1, -1, 1) x + 2 >= 0,
+         An inequality (1, 1, 1, -1) x + 2 >= 0)
+        sage: v = vector(ZZ, [1,1,-1])
+        sage: L = IntegerLattice([v])
+        sage: C = L.voronoi_cell()
+        sage: C.Hrepresentation()
+        (An inequality (-2, -2, 2) x + 3 >= 0, An inequality (2, 2, -2) x + 3 >= 0)
+        sage: C.Vrepresentation()
+        (A line in the direction (0, 1, 1),
+         A line in the direction (1, 0, 1),
+         A vertex at (0, 0, -3/2),
+         A vertex at (0, 0, 3/2))
+
+    Verify that :issue:`37086` is fixed::
+
+        sage: from sage.modules.free_module_integer import IntegerLattice
+        sage: l  = [7, 0, -1, -2, -1, -2, 7, -2, 0, 0, -2, 0, 7, -2, 0, -1, -2, -1, 7, 0 , -1, -1, 0, -2, 7]
+        sage: M = matrix(5, 5, l)
+        sage: c = IntegerLattice(M).voronoi_cell(radius=35)
+        sage: M.det()
+        12858
+        sage: c.volume()
+        12858
     """
     dim = basis.dimensions()
+    # LLL-reduce for efficiency.
+    basis = basis.LLL()
+    if radius is None:
+        from sage.rings.real_double import RDF
+        # Convert the basis matrix to use RDF numbers for efficiency when we
+        # perform the QR decomposition.
+        tranposed_RDF_matrix = (basis.transpose()).change_ring(RDF)
+        # We then perform the QR decomposition.
+        R = tranposed_RDF_matrix.QR()[1]
+        # The radius is then an upper bound for the covering radius.
+        radius = sum(R[i,i]**2 for i in range(dim[0]))
+        # We then divide by 4 as we will divide the basis by 2 later on.
+        radius = radius / 4
     artificial_length = None
     if dim[0] < dim[1]:
-        # introduce "artificial" basis points (representing infinity)
-        def approx_norm(v):
-            r, r1 = (v.inner_product(v)).sqrtrem()
-            return r + (r1 > 0)
-        artificial_length = max(approx_norm(v) for v in basis) * 2
-        additional_vectors = identity_matrix(dim[1]) * artificial_length
+        from sage.rings.rational_field import QQ
+        # Introduce "artificial" basis points (representing infinity).
+        additional_vectors = (QQ**dim[1]).subspace(basis).complement().basis()
+        # We then make the artificial points integers.
+        for v in additional_vectors:
+            v *= v.denominator()
+        additional_vectors = matrix(additional_vectors)
+        from sage.rings.rational_field import ZZ
+        additional_vectors = additional_vectors.change_ring(ZZ)
```

Comment:
```suggestion
        additional_vectors = matrix(additional_vectors)
```
```suggestion
 ```
I think matrices over `QQ` have an `LLL` method, so there should be no need to convert to integers. Also, trying to do this could cause an error if the base ring is not `ZZ` or `QQ`.